CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes

veronikaslc opened the pull request:
To test **on the CARDS-2333-test branch**:

- build and start in prems mode
- create a new form, leave it in edit mode, check (in a different tab) that after two minutes the form is checked in
- save & view, edit again, wait one minute, make sure the form isn't checked in, make a change and save&continue editing, wait one minute, make sure the form isn't checked in, wait another minute, make sure the form is checked in
- generate a Visit Information form, wait 2 minutes, make sure the Survey Status form is checked in; fill the survey as a patient, submit, make sure the Survey Status form was updated to include the Survey Submited date.

veronikaslc committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/StaleFormsCheckinScheduler.java:
```diff
@@ -0,0 +1,75 @@
+    {
+        try {
+            // Every 5 minutes
+            final ScheduleOptions options = this.scheduler.EXPR("*/5 * * * *");
```
Isn't that every 5 seconds?

veronikaslc replied:
Took the idea from [here](https://crontab.guru/every-5-minutes)

sdumitriu replied:
Oh, the java (quartz) and crontab formats are slightly different. I use cronmaker.com, and it suggests this:

`0 0/5 * 1/1 * ? *`

veronikaslc replied:
fixed

veronikaslc committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Fixed job schedule frequency

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Fixed since version

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Better API usage

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
More readable code

marta- commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/internal/DraftsAnswersCleanupTask.java:
```diff
@@ -103,7 +103,7 @@ public void run()
                     // belonging to a visit
                     + "  inner join [cards:Form] as visitInformation on visitInformation.subject = dataForm.subject"
-                    + "    inner join [cards:BooleanAnswer] as submitted on submitted.form = visitInformation.[jcr:uuid]"
+                    + "  inner join [cards:BooleanAnswer] as submitted on submitted.form = visitInformation.[jcr:uuid]"
```
Let's drop this change. It's unrelated and the indentation of the second "inner join" actually improves readability.

sdumitriu replied:
Compilation fails without this change. It was actually caused by merging another PR which somehow wasn't built. I thought I did build before putting up the PR...

marta- replied:
Now fixed in dev https://github.com/data-team-uhn/cards/pull/1591 please rebase and remove this change

sdumitriu replied:
Done.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/StaleFormsCheckinTask.java:
```diff
@@ -73,6 +73,7 @@ public void run()
             .getServiceResourceResolver(Map.of(ResourceResolverFactory.SUBSERVICE, "staleFormsCheckin"))) {
             this.rrp.push(resolver);
             mustPopResolver = true;
+            final VersionManager vm = resolver.adaptTo(Session.class).getWorkspace().getVersionManager();
```
I'm not a fan of these extra-brief variable names that require you to look at their definition to see what they're about. Can we call it `versionManager` instead?

sdumitriu replied:
Sure.

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Fix service user setup

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Switch to using a special jcr:lastCheckedOut property since jcr:lastModified doesn't get updated when the form is checked out

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Remove unneeded parameter

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Better variable name

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Use both lastModified and lastCheckedOut to check if a form is stale

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Move the autocheckin code in the cards-versioning module

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Make sure the Survey Status form is checked out before it is updated, now that we automatically check in forms

sdumitriu committed:
CARDS-2333: Automatically check in forms that haven't been modified in more than 30 minutes
Also index the jcr:isCheckedOut property for faster searches

sashaandjic reviewed (approved):
tested
